ext/standard/pack.c

```diff
@@ -1003,108 +989,78 @@ PHP_FUNCTION(unpack)
 						add_assoc_long(return_value, n, v);
 						break;
 					}
 
-					case 'i':
-					case 'I': {
-						zend_long v;
-						int issigned = 0;
-
-						if (type == 'i') {
-							issigned = input[inputpos + (machine_little_endian ? (sizeof(int) - 1) : 0)] & 0x80;
-						}
-
-						v = php_unpack(&input[inputpos], sizeof(int), issigned, int_map);
+					case 'i':   /* signed integer, machine size, machine endian */
+					case 'I': { /* unsigned integer, machine size, machine endian */
+						unsigned int x;
+						memcpy(&x, &input[inputpos], sizeof(unsigned int));
+						zend_long v = (type == 'i') ? (int) x : x;
 						add_assoc_long(return_value, n, v);
 						break;
 					}
 
-					case 'l':
-					case 'L':
-					case 'N':
-					case 'V': {
-						int issigned = 0;
-						int *map = machine_endian_long_map;
+					case 'l':   /* signed machine endian   */
+					case 'L':   /* unsigned machine endian */
+					case 'N':   /* unsigned big endian     */
+					case 'V': { /* unsigned little endian  */
 						zend_long v = 0;
+						uint32_t x;
+						memcpy(&x, &input[inputpos], 4);
 
-						if (type == 'l' || type == 'L') {
-							issigned = input[inputpos + (machine_little_endian ? 3 : 0)] & 0x80;
-						} else if (type == 'N') {
-							issigned = input[inputpos] & 0x80;
-							map = big_endian_long_map;
-						} else if (type == 'V') {
-							issigned = input[inputpos + 3] & 0x80;
-							map = little_endian_long_map;
-						}
-
-						if (SIZEOF_ZEND_LONG > 4 && issigned) {
-							v = ~INT_MAX;
+						if (type == 'l') {
+							v = (int32_t) x;
+						} else if ((type == 'N' && MACHINE_LITTLE_ENDIAN) || (type == 'V' && !MACHINE_LITTLE_ENDIAN)) {
+							v = php_pack_reverse_int32(x);
+						} else {
+							v = x;
 						}
 
-						v |= php_unpack(&input[inputpos], 4, issigned, map);
-						if (SIZEOF_ZEND_LONG > 4) {
- 							if (type == 'l') {
-								v = (signed int) v;
-							} else {
-								v = (unsigned int) v;
-							}
-						}
 						add_assoc_long(return_value, n, v);
 						break;
 					}
 
 #if SIZEOF_ZEND_LONG > 4
-					case 'q':
-					case 'Q':
-					case 'J':
-					case 'P': {
-						int issigned = 0;
-						int *map = machine_endian_longlong_map;
+					case 'q':   /* signed machine endian   */
+					case 'Q':   /* unsigned machine endian */
+					case 'J':   /* unsigned big endian     */
+					case 'P': { /* unsigned little endian  */
 						zend_long v = 0;
-
-						if (type == 'q' || type == 'Q') {
-							issigned = input[inputpos + (machine_little_endian ? 7 : 0)] & 0x80;
-						} else if (type == 'J') {
-							issigned = input[inputpos] & 0x80;
-							map = big_endian_longlong_map;
-						} else if (type == 'P') {
-							issigned = input[inputpos + 7] & 0x80;
-							map = little_endian_longlong_map;
-						}
-
-						v = php_unpack(&input[inputpos], 8, issigned, map);
+						uint64_t x;
+						memcpy(&x, &input[inputpos], 8);
 
 						if (type == 'q') {
-							v = (zend_long) v;
+							v = (int64_t) x;
```

Comment:
Far from being a C expert but isn't the following:
```suggestion
							v = *(int64_t*) &x;
```
the correct way of type-punning? As IIRC signed/unsigned casts are UB if the value exceeds the range of a type.

Reply:
For casts between integers, the existing code is fine. What you may have in mind is a bitwise cast between float and int.

Reply:
I'm not sure, I double checked the standard and apparently unsigned to signed is UB when the value doesn't fit, but almost all compiler do two complement (according to https://unspecified.wordpress.com/category/c-3/ and if I read this correctly).

So should only be an issue on unicorn compilers

Reply:
The standard (6.3.1.3 §3) mentions this conversion produce a result which is implementation-defined, not UB.
Is it OK to keep it like this?